serializer_impl.hh

```diff
@@ -782,4 +782,138 @@ unknown_variant_type deserialize(Input& in, boost::type<unknown_variant_type>) {
         return unknown_variant_type{ index, std::move(v) };
     });
 }
+
+template <typename T>
+class range_deserializer {
+    utils::input_stream _in;
+    uint32_t _size;
+
+public:
+    range_deserializer(const utils::input_stream& in)
+        : _in(in)
+    {
+        _size = deserialize(_in, boost::type<uint32_t>());
+    }
+
+    // Get the number of items in the vector
+    size_t size() const noexcept {
+        return _size;
+    }
+
+    class iterator;
+
+    // Sentinel for the end() position
+    class sentinel {
+        uint32_t _idx;
+
+        sentinel(uint32_t idx) noexcept
+            : _idx(idx)
+        { }
+
+        friend class range_deserializer;
+    public:
+        ssize_t operator-(const iterator& it) const noexcept {
+            return _idx - it._idx;
+        }
+    };
+
+    // Forward iterator
+    class iterator {
+        utils::input_stream _in;
+        uint32_t _idx = 0;
+        std::optional<T> _item;
+
+        iterator(const utils::input_stream& in, uint32_t idx, std::optional<T>&& item)
+            : _in(in)
+            , _idx(idx)
+            , _item(std::move(item))
+        { }
+
+        friend class range_deserializer;
+    protected:
+        iterator(const utils::input_stream& in)
+            : _in(in)
+            , _idx(0)
+        { }
+
+    private:
+        static size_type read_frame_size(utils::input_stream& in) {
+            auto sz = deserialize(in, boost::type<size_type>());
+            if (sz < sizeof(size_type)) {
+                throw std::runtime_error(fmt::format("IDL frame truncated: expected to have at least {} bytes, got {}", sizeof(size_type), sz));
+            }
+            return sz - sizeof(size_type);
+        }
+
+   public:
+        iterator() = default;
+
+        iterator(iterator&& o) noexcept
+            : _in(std::move(o._in))
+            , _idx(o._idx)
+            , _item(std::exchange(o._item, std::nullopt))
+        { }
+
+        iterator(const iterator& o) = default;
+
+        iterator& operator=(iterator&&) = default;
+        iterator& operator=(const iterator&) = default;
+
+        bool operator==(const sentinel& s) noexcept {
+            return _idx >= s._idx;
+        }
+
+        bool operator==(const iterator& it) noexcept {
+            return _idx == it._idx;
+        }
+
+        // Return const& as the input stream cannot be modified
+        const T& operator*() {
```

Comment:
const

Reply:
It can't be const since it engages _item.

Reply:
You can make `_item` `mutable`. It is only a cache after all. I think the interface's const correctness should reflect the logical state of the object, rather then such implementation details.

Reply:
return `T&` rather than `const T&` is not really required.
The root problem of defining this method as `const T& operator*() const` is that
it engages `this->_item` and therefore it modifies the iterator.

Reply:
Like I said, you can make `this->_item` mutable to work around this problem. Such cache-only members are a typical use-case for mutable. The iterator is mutated logically when it is incremented. Dereferencing doesn't modify its state.

Reply:
input iterators should return T, not const T&.

Reply:
I agree in theory, but it looks like the standard requires a reference, see:
https://en.cppreference.com/w/cpp/iterator/input_iterator
and https://en.cppreference.com/w/cpp/iterator/indirectly_readable

But anyhow, we'd need to keep the item somewhere for `operator->`, no?

Reply:
I think input iterators typically don't support `operator->`.

Reply:
The `LegacyInputIterator` concept requires `operator->()`. See the table of expressions that are required to be valid at https://en.cppreference.com/w/cpp/named_req/InputIterator.

Reply:
But https://en.cppreference.com/w/cpp/iterator/input_iterator doesn't.
It requires only `*it` (via https://en.cppreference.com/w/cpp/iterator/indirectly_readable)

I'm working on a version returning the item for `operator*` and it is much simpler and elegant.